### PR TITLE
Hoist and distribute type parameter constraints over type parameters …

### LINKED ISSUE #33340: Value of type T is not assignable to T & U even when T extends U
At least in the case of `U` being a nonempty finite union of singleton types, it seems that `T` extending `U` does not imply that `T` is assignable to `T & U`, even though every value of type `T` is also a value of type `U`.

However, the distributive conditional type `T extends U ? T : never` is considered assignable to `U` even though it should have the same set of values as `T` since we assume that `T` extends `U`.

**TypeScript Version:**  3.5.1 and 3.7.0-dev.20190907

**Search Terms:** enum intersection extends not reduced

**Code**

```ts
type A = 1 | 2;
function f<T extends A>(a: T): A & T  { return a; } // Error: Type '1' is not assignable to type '2'.
function g<T extends A>(a: T extends A ? T : never): A & T  { return a; } // OK
```

**Expected behavior:** Compiles.

**Actual behavior:** Does not compile.

**Playground Link:** https://www.typescriptlang.org/play/index.html#code/C4TwDgpgBAglC8UCMUA+UBMBuAUAMwFcA7AY2AEsB7IqPAHgBUoIAPYCIgEwGdYA+ABQBDAFxQGASjFwAZOKhQA3lABOEYARU0hWKAF8gA

**Related Issues:** https://github.com/microsoft/TypeScript/issues/28830

For context, I am attempting to write a function that takes a union tag and a payload and returns a matching new instance of one of the members of a tagged union.

Here is some further exploration in that direction: https://www.typescriptlang.org/play/index.html#code/C4TwDgpgBAglC8UCMUA+UBMBuAUAMwFcA7AY2AEsB7IqPAHgBUoIAPYCIgEwGdYA+ABQBDAFxQGASjFwAZOKhQA3lABOEYARU0hWKAF98xMlRoBzRszYce-YWKat2XXnAD88sUQgA3CCqmwUHJMSqrqmtq6BjjkROwqeEIk0ACySjgKCqLIuAax8YnJUABy6ZlQ2dg40YSkFNRQABYWjtYugtmSYgIpqMUSQUqdegrKahpaQ3pRhnUmUOQtVs62nZZONm4eUF6+-t29-YOKw6NhEzQn0-pQs8YNAFZLG+12Q-YjrSuHUO5Mnj4-AEen0BnITh8zuMIlMZrV7jQANbPNqBDr2daorb-HaA-bvcSfZY2EFHdwQhgjcGnAF7ULQyZXOFGeo0AA2KJWMHRBMpmJWpLBvJGv22uyBB1Bx1O9PCjKE1z0QA

### PR DESCRIPTION
**Pull request by weswigham**
…when comparing against union targets when fetching union constraints

Fixes #33340

The `error` baseline changes are because I disabled the redundant error info pyramids from the distributive constraint error - I'd only really want to add them back in if I changed them to dedent like overload errors, since as-is they didn't really flow.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -12954,10 +12954,18 @@ namespace ts {
...
-                            errorInfo = saveErrorInfo;
+                    // This also handles type parameters, as a type parameter with a union constraint compared against a union
+                    // needs to have its constraint hoisted into an intersection with said type parameter, this way
+                    // the type param can be compared with itself in the target (with the influence of its constraint to match other parts)
+                    // For example, if `T extends 1 | 2` and `U extends 2 | 3` and we compare `T & U` to `T & U & (1 | 2 | 3)`
+                    const constraint = getUnionConstraintOfIntersection(source.flags & TypeFlags.Intersection ? (<IntersectionType>source).types: [source], !!(target.flags & TypeFlags.Union));
+                    if (constraint && (source.flags & TypeFlags.Intersection || target.flags & TypeFlags.Union)) {
+                        const filtered = filterType(constraint, c => c !== source);
+                        if (filtered === constraint) { // Skip comparison if expansion contains the source itself
```
Comment (jack-williams) [inline src/compiler/checker.ts]: Is `everyType(candidate, c => c !== source)` clearer?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -7967,7 +7967,7 @@ namespace ts {
                         constraint = getConstraintOfType(constraint);
                     }
                     if (constraint) {
-                        constraints = append(constraints, constraint);
+                        constraints = append(constraints, targetIsUnion ? intersectTypes(constraint, t) : constraint);
```
[the remark this answers is not in the document]
Comment (weswigham) [reply src/compiler/checker.ts]: Yeah, that'd be... the same.

File: src/compiler/checker.ts (hunk 1 above)
Comment (weswigham) [reply src/compiler/checker.ts]: Sure.

**Comment by weswigham** [comment]
@typescript-bot perf test this

**Comment by weswigham** [comment]
@ahejlsberg perf here is fine, any other comments, questions, or concerns?

**Comment by weswigham** [comment]
Just because while I feel it's probably OK, we should probably do this on relationship/inference changes for diligence (and I don't want to deal with breaks tomorrow if there are any, and would rather deal with them now):

@typescript-bot run dt
@typescript-bot user test this
@typescript-bot test this

**Comment by weswigham** [comment]
Excellent, DT is clean and the others are expected given the last merge date on this PR.

**Pull request merged**